### PR TITLE
x509pop server plugin support for servers trust bundle

### PR DESCRIPTION
**Pull request by kfox1111**
Enables the x509pop node attestor server plugin to be configured to use the SPIRE Servers own trust bundle.

### CODE REVIEW

File: pkg/server/plugin/nodeattestor/x509pop/x509pop.go
Comment (azdagron) [inline pkg/server/plugin/nodeattestor/x509pop/x509pop.go:125]: Rather than kick off a goroutine that may or may not be run after the IdentityProvider host service is ready, it would be better to lazily initialize this in the first call to Attest.

**Comment by kfox1111** [comment]
> Using the bundle, without any sort of restriction, seems scary. Doesn't this imply that any workload can just turn around and attest as a node?

Yeah, that I think can be handled like:
```
    NodeAttestor "x509pop" {
      plugin_data {
        spire_trust_bundle = true
        # Only allow spiffe identifiers of the form spiffe://<trustDomain>/k8s-spire-agent-helper/<hostname>
        # Generate identifiers that look like spiffe://<trustDomain>/x509pop/k8s-spire-agent/<hostname>
        agent_path_template = "{{ ${d}p := printf \"spiffe://%s/k8s-spire-agent-helper/\" .TrustBundle }}{{ ${d}s := printf \"%s\" (index .Certificate.URIs 0) }}{{ if gt (len ${d}s) (len ${d}p) }}{{ ${d}ps := slice ${d}s 0 (len ${d}p)
 }}{{ if eq ${d}ps ${d}p }}{{ printf \"/x509pop/k8s-spire-agent/%s\" (slice ${d}s (len ${d}p)) }}{{ end }}{{ end }}"
      }
    }
```

if it agent_path renders out to "", then it should be disallowed.

I did hit:
* https://github.com/spiffe/spire/issues/5574
* https://github.com/spiffe/spire/issues/5573

Trying to get it to work though.

File: pkg/server/plugin/nodeattestor/x509pop/x509pop.go
Comment (kfox1111) [reply pkg/server/plugin/nodeattestor/x509pop/x509pop.go]: Done

**Comment by azdagron** [comment]
Hey @kfox1111, do you have an arch diagram or something that spells out the use case clearly that you can share (either here or privately in slack with the maintainer group)? Before we consider taking this we'd like to see if there are alternatives.

**Comment by kfox1111** [comment]
Here's what I have done / working on doing. Got the host -> k8s part working. Still working on the k8s -> vm part.
x509pop

**Comment by kfox1111** [comment]
Related to: https://github.com/spiffe/spire/issues/5206

**Comment by kfox1111** [comment]
Used with the sprig pr, this works:

        agent_path_template = "{{ $$p := printf \"spiffe://%s/k8s-spire-agent-helper/\" .TrustDomain }}{{ $$s := printf \"%s\" (index .Certificate.URIs 0) }}{{ if hasPrefix $$p $$s }}{{ printf \"/x509pop/k8s-spire-agent/%s\" (trimPrefix $$p $$s) }}{{ else }}{{ fail \"Invalid prefix\" }}{{ end }}"

**Comment by kfox1111** [comment]
Updated diagram:
x509pop

**Comment by kfox1111** [comment]
@sorindumitru I think we're pretty much on the same page.

We could make a new x509svidpop plugin but it would end up being almost the same code as the x509pop one. (The patch here is pretty small). So a lot more stuff to maintain.

the svid use case really isn't that different from the normal x509pop one. Either way you probably want some way to filter out which certs are allowed to a subset of all possible ones. That can be done via agent_path_template as done in this pr without any special code in the plugin.

100% agree on the future ability to use some other spire trustDomain too. Was going to propose that in a different patch. The primary use case for me being:

k8s multitenant cluster. Secured with spire. Call this, the resource provider cluster.

Tenant uses resource provider cluster with kubevirt to launch a bunch of vm's for their tenant. Inside, they launch their own spire server, and workloads (maybe even their own k8s cluster in the vms). It would greatly simplify management if they could use x509 certs from the resource provider cluster to node attest to their own spire-server.

That again, would basically be this same x509pop plugin as described here, but getting its trustBundle from disk and keep it refreshed (existing functionality extended to refresh from disk), or get it over vsock (not sure which is best yet)

**Comment by sorindumitru** [comment]
I agree it's going to be very similar to `x509pop`, but dealing with that is an implementation detail, I'm sure some parts can be shared. I just don't think that the template language is a good way of dealing with those differences.

File: doc/plugin_server_nodeattestor_x509pop.md
Comment (azdagron) [inline doc/plugin_server_nodeattestor_x509pop.md:22]: Maybe `svid_path_prefix`? Also...
```suggestion
| `svid_prefix`            | The prefix of the SVID to use for matching valid SVIDS and exchanging them for Node SVIDs                                                                                                                                                    | /spire-exchange                         |
```

File: doc/plugin_server_nodeattestor_x509pop.md
Comment (azdagron) [inline doc/plugin_server_nodeattestor_x509pop.md:25]: nit: use a section link to the correct section (https://docs.github.com/en/get-started/writing-on-github/getting-started-with-writing-and-formatting-on-github/basic-writing-and-formatting-syntax#section-links)

File: doc/plugin_server_nodeattestor_x509pop.md
Comment (azdagron) [inline doc/plugin_server_nodeattestor_x509pop.md:48]:
```suggestion
## SVID Path Prefix
```

File: doc/plugin_server_nodeattestor_x509pop.md
Comment (azdagron) [inline doc/plugin_server_nodeattestor_x509pop.md:80]: Maybe `SVIDPathSuffix` better implies that a prefix has been stripped?

File: pkg/server/plugin/nodeattestor/x509pop/x509pop.go
Comment (azdagron) [inline pkg/server/plugin/nodeattestor/x509pop/x509pop.go:42]: SVIDPrefix? Also, if we don't need to distinguish between this flag being unset and "" (e.g. if "" can safely imply unset), I'd suggest this be of type `string` to reduce pointery-ness.

File: pkg/server/plugin/nodeattestor/x509pop/x509pop.go
Comment (azdagron) [inline pkg/server/plugin/nodeattestor/x509pop/x509pop.go:50]: To be consistent, unexport these.

File: pkg/server/plugin/nodeattestor/x509pop/x509pop.go
Comment (azdagron) [inline pkg/server/plugin/nodeattestor/x509pop/x509pop.go:99]: local variable naming conventions are `lowerCamel`
```suggestion
	var spiffePrefix string
```

File: pkg/server/plugin/nodeattestor/x509pop/x509pop.go
Comment (azdagron) [inline pkg/server/plugin/nodeattestor/x509pop/x509pop.go:231]:
```suggestion
	svidPath := ""
```

File: pkg/server/plugin/nodeattestor/x509pop/x509pop.go
Comment (azdagron) [inline pkg/server/plugin/nodeattestor/x509pop/x509pop.go:233]: nit: if you invert the condition, you can do the early failure return when there are no URIs and then not need an else (reduces indentation)

File: pkg/server/plugin/nodeattestor/x509pop/x509pop.go
Comment (azdagron) [inline pkg/server/plugin/nodeattestor/x509pop/x509pop.go:64]: Should we be doing sanity checking that folks aren't using both "spire_trust_bundle" and the other ca related flags?

**Review by azdagron — COMMENTED** [review]
Thanks for this, @kfox1111. The new design eliminates most of my concerns.

I'd like to explore names other than `spire_trust_bundle` and the default `/spire-exchange` prefix value.... I think we can find something a little clearer....

Also, there are no tests for the new functionality. We'll need those before we can merge.

File: doc/plugin_server_nodeattestor_x509pop.md
Comment (kfox1111) [inline doc/plugin_server_nodeattestor_x509pop.md:25]:
```suggestion
| `agent_path_template` | A URL path portion format of Agent's SPIFFE ID. Describe in text/template format.                                                                                                                                                              | `See [Agent Path Template](#agent-path-template) for details`   |
```

File: pkg/server/plugin/nodeattestor/x509pop/x509pop.go
Comment (MarcosDY) [inline pkg/server/plugin/nodeattestor/x509pop/x509pop.go:73]: is not the same error happens in line 80?
may we return here instead?

File: pkg/server/plugin/nodeattestor/x509pop/x509pop.go
Comment (MarcosDY) [inline pkg/server/plugin/nodeattestor/x509pop/x509pop.go:117]: this can be simplified:
```suggestion
	svidPrefix :=  "/spire-exchange/"
	if hclConfig.SVIDPrefix != nil {
		svidPrefix = *hclConfig.SVIDPrefix
		if !strings.HasSuffix(svidPrefix, "/") {
			svidPrefix += "/"
		}
	}
```

File: pkg/server/plugin/nodeattestor/x509pop/x509pop.go
Comment (MarcosDY) [inline pkg/server/plugin/nodeattestor/x509pop/x509pop.go:307]: if this happens we'll reject attestations, but it is going to be hard to understand why it is happening (basically we are not going to have a trusted bundle)
May we at least return wartning or error?

File: pkg/server/plugin/nodeattestor/x509pop/x509pop_test.go
Comment (MarcosDY) [inline pkg/server/plugin/nodeattestor/x509pop/x509pop_test.go:163]: can you add test case where no trustedbundle is returned? 
and how about defaults?
are

File: pkg/server/plugin/nodeattestor/x509pop/x509pop.go
Comment (kfox1111) [reply pkg/server/plugin/nodeattestor/x509pop/x509pop.go:73]: Hmm.... The check here checks if both are configured. The check at 80 ensures at least one is configured. Both are needed I think.

I was thinking status.ReportError was fatal. Is it not? Maybe all the ReportError's here then need retruns?

File: pkg/server/plugin/nodeattestor/x509pop/x509pop.go
Comment (kfox1111) [reply pkg/server/plugin/nodeattestor/x509pop/x509pop.go:305]: Maybe we log a warning about not currently finding a trust bundle? Its nor fatal as it might show up at some point later?

File: pkg/server/plugin/nodeattestor/x509pop/x509pop.go
Comment (MarcosDY) [reply pkg/server/plugin/nodeattestor/x509pop/x509pop.go:73]: mmm, it is like if both are configured, you must never get the second one right?
status.Report is not a fatal, but just adding a list of errors, than then are used in configure or verify,
but this approach has a good win than you can get the complete list of errors to resolve all together,
but it makes very hard to follow and see dependency errors...
In any case we are fine for now

File: pkg/server/plugin/nodeattestor/x509pop/x509pop.go
Comment (MarcosDY) [reply pkg/server/plugin/nodeattestor/x509pop/x509pop.go:305]: yeap, I agree but this attestation will fail and mention a verification error
so a warning may help users to understand than that the that attestation failure is cuased by the attestor being unabled to get the trust bundle.

Comment (kfox1111) [reply pkg/server/plugin/nodeattestor/x509pop/x509pop.go:314]: Added a warning.

File: pkg/server/plugin/nodeattestor/x509pop/x509pop_test.go
Comment (kfox1111) [reply pkg/server/plugin/nodeattestor/x509pop/x509pop_test.go:163]: Struggling a bit with how to write the tests....

Seems like I need to create a new cert in ./test/fixture/nodeattestor/x509pop/generate.go with the url based spiffe id to test with? There are some magic numbers in that file. Are they just arbitrary, or are there some reason for the specific numbers chosen?

Comment (kfox1111) [reply pkg/server/plugin/nodeattestor/x509pop/x509pop_test.go:163]: Reading through it some more, it looks like they just have patterns in it. so rather arbitrary.

Comment (kfox1111) [reply pkg/server/plugin/nodeattestor/x509pop/x509pop_test.go:208]: Ok. I have tests for no trust bundle, along with a valid svid, an invalid svid, and a no svid.

**Review by MarcosDY — APPROVED** [review]

**Pull request merged**
